File: android/src/main/kotlin/cat/bcn/commonmodule/flutter/osam_common_module_flutter/messaging/MessagingBridge.kt
package cat.bcn.commonmodule.flutter.osam_common_module_flutter.messaging

import android.content.Context
import cat.bcn.commonmodule.messaging.MessagingWrapper
import io.flutter.plugin.common.EventChannel
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

class MessagingBridge(val context: () -> Context?) : EventChannel.StreamHandler, MessagingWrapper {

    companion object {
        private const val TOPIC = "topic"
        private const val ACTION = "action"
        private const val SUBSCRIBE = "subscribe"
        private const val UNSUBSCRIBE = "unsubscribe"
        private const val GET_TOKEN = "getToken"
    }

    private var eventSink: EventChannel.EventSink? = null

    override suspend fun subscribeToTopic(topic: String) {
        // We must switch to the Main Thread to communicate with Flutter
        withContext(Dispatchers.Main) {
            eventSink?.success(
                mapOf(
                    TOPIC to topic,
                    ACTION to SUBSCRIBE
                )
            )
            Unit
        }
    }

    override suspend fun unsubscribeFromTopic(topic: String) {
        // We must switch to the Main Thread to communicate with Flutter
        withContext(Dispatchers.Main) {
            eventSink?.success(
                mapOf(
                    TOPIC to topic,
                    ACTION to UNSUBSCRIBE
                )
            )
            Unit
        }
    }

    override suspend fun getToken(): String {
        val prefs = context()?.getSharedPreferences("FlutterSharedPreferences", Context.MODE_PRIVATE)
        withContext(Dispatchers.Main) {
            eventSink?.success(
                mapOf(
                    TOPIC to "",
                    ACTION to GET_TOKEN
                )
            )
            Unit
        }
        return prefs?.getString("flutter.fcm_token", "") ?: ""
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }
}

File: ios/Classes/messaging/MessagingBridge.swift
//
//  MessagingBridge.swift
//  
//
//  Created by Gustavo Lorena on 26/11/25.
//
import OSAMCommon
import Flutter

class MessagingBridge: NSObject, FlutterStreamHandler, MessagingWrapper {

    private var eventSink: FlutterEventSink?
    private static let TOPIC = "topic"
    private static let ACTION = "action"
    private static let GET_TOKEN = "getToken"
    private static let SUBSCRIBE = "subscribe"
    private static let UNSUBSCRIBE = "unsubscribe"

    func subscribeToTopic(topic: String) async {
        // Platform channel communication must happen on the main thread.
        DispatchQueue.main.async { [weak self] in
            self?.eventSink?([
                                 MessagingBridge.TOPIC: topic,
                                 MessagingBridge.ACTION: MessagingBridge.SUBSCRIBE
                             ])
        }
    }

    func unsubscribeFromTopic(topic: String) async {
        // Platform channel communication must happen on the main thread.
        DispatchQueue.main.async { [weak self] in
            self?.eventSink?([
                                 MessagingBridge.TOPIC: topic,
                                 MessagingBridge.ACTION: MessagingBridge.UNSUBSCRIBE
                             ])
        }
    }

    func getToken() async -> String {
        // Asynchronously send a notification event over the EventChannel.
        // This is a "fire-and-forget" operation; the function does not wait for it to complete.
        DispatchQueue.main.async { [weak self] in
            self?.eventSink?([
                                 MessagingBridge.TOPIC: "",
                                 MessagingBridge.ACTION: MessagingBridge.GET_TOKEN
                             ])
        }

        // Immediately read the token from UserDefaults.
        // The "flutter." prefix is automatically added by Flutter's shared_preferences
        // plugin and must be included when reading the value natively.
        let token = UserDefaults.standard.string(forKey: "flutter.fcm_token")

        // Return the retrieved token, or an empty string if it was nil.
        return token ?? ""
    }
    

    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self.eventSink = nil
        return nil
    }
}
